### ISSUE TITLE
ServiceAccount of jfs-mount pods

### ISSUE THREAD

**Issue by Iced-Sun** [juicedata/juicefs-csi-driver#167]
We use [IAM roles for service accounts ](https://docs.aws.amazon.com/eks/latest/userguide/create-service-account-iam-policy-and-role.html) to grant access to S3 buckets. Both juicefs-csi-controller & juicefs-csi-node work fine with manually mounting the S3 backend storages.

The problem is that the jfs-mount pods are provisioned with the default service-account, which doesn't have access to the S3 backend, leading to mounted but unusable pv/pvc. 

Maybe more configurations of jfs-mount could be exposed in `.values.storageClasses.mountPod.*`?

BTW, what is the minimal S3 privilidges to make juicefs work?

**Comment by davies** [comment]
@Iced-Sun Do you have some hint on how to select the service-account? My understanding is that all the container uses the same role to access S3. We will try to reproduce this issue, and get back to you.

The mount options will addressed by #181 hopefully.

The minimal S3 privilidges are: PUT, GET, DELETE on objects with prefix as volume name. When the volume is not formatted before, CREATE bucket is also needed.

Sorry for the late reply.

**Comment by Iced-Sun** [comment]
Apology for the late reply.

TL;DR 
Make mountPod inherit the service account of juicefs-csi-node.

-------

> The mount options will addressed by #181 hopefully.
> 
It seems that the mount options don't affect the S3 crendentials, hence it may not help much.

AFAIK, to make the mountPods (pods with name `juicefs-<node-name>-<pvc-name>`) access to the S3 backends, one should either 
1. provide the access key & secret key
or
2. link the instance role to the k8s worknode

Both the two options work, with some concerns. Option 1 doesn't work well with key rotation/revocation; option 2 makes all the pods gain the access to the underlying S3 bucket.

A common approach for access controll of EKS would be utilizing the annotation `eks.amazonaws.com/role-arn` of a serviceAccount, c.f.
1. https://github.com/kubernetes-sigs/external-dns/blob/master/docs/tutorials/aws.md#amazon-eks
2. https://docs.aws.amazon.com/zh_cn/eks/latest/userguide/aws-load-balancer-controller.html

For reference, here are the configurations currently adopted by our EKS deployment:

1. create an IAM OIDC provider for the EKS cluster https://docs.amazonaws.cn/en_us/eks/latest/userguide/enable-iam-roles-for-service-accounts.html

2. create IAM policy to enable access to the S3 bucket
```js
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Action": [
                "s3:*"
            ],
            "Resource": [
                "arn:aws-cn:s3:::<REDACTED>/*",
                "arn:aws-cn:s3:::<REDACTED>"
            ]
        }
    ]
}
```

3. create IAM role for EKS service account, with the policy created in 2 and the following assume role policy:
```js
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Principal": {
        "Federated": "arn:aws-cn:iam::<REDACTED>:oidc-provider/oidc.eks.cn-northwest-1.amazonaws.com.cn/id/<REDACTED>"
      },
      "Action": "sts:AssumeRoleWithWebIdentity",
      "Condition": {
        "StringEquals": {
          "oidc.eks.cn-northwest-1.amazonaws.com.cn/id/<REDACTED>:sub": [
            "system:serviceaccount:juicefs:juicefs-csi-controller-sa",
            "system:serviceaccount:juicefs:juicefs-csi-node-sa",
            "system:serviceaccount:<NAMESPACE>:<SERVICE-ACCOUNT-NAME>"
          ]
        }
      }
    }
  ]
}
```

4. within the namespace `juicefs`, create the service accounts of `juicefs-csi-controller-sa` and `juicefs-csi-node-sa`, with configuration:
```yaml
apiVersion: v1
kind: ServiceAccount
metadata:
  annotations:
    eks.amazonaws.com/role-arn: arn:aws-cn:iam::<REDACTED>:role/<REDACETD>
....
```

5. within the namespace `juicefs`, deploy the juicefs-csi-controller with service account `juicefs-csi-controller-sa` and the juicefs-csi-node with service account `juicefs-csi-node-sa`

6. create a k8s Secret, in which the access key is not required:
```yaml
apiVersion: v1
stringData:
  bucket: <REDACTED>
  metaurl: <REDACTED>
  name: juicefs-data
  storage: s3
kind: Secret
metadata:
  name: juicefs-data-secret
  namespace: juicefs
type: Opaque
```

------

Now let's apply https://github.com/juicedata/juicefs-csi-driver/blob/master/examples/basic/basic.yaml with proper modifications. 

We could observe the following:
1. volume is successfully formatted and mounted, since no S3 access is required
2. `/data/out.txt` presents, since no S3 access is required
3. the underlying S3 bucket is empty, even after a while and data is expected (through `>>/data/out.txt`)
4. `cat /data/out.txt` hangs

Obviously, the cause of 3, 4 is the lack of S3 access. One can verify that by doing the following:
1. delete the created pod, pv and pvc
2. shell into `juicefs-csi-controller` or `juicefs-csi-node`
3. execute `juicefs mount <meta_url> /mnt/temp`
4. execute `echo sometext > /mnt/temp/some.txt` 

Everything should work fine (object presents in S3 bucket; `/mnt/temp/some.txt` has contents).

------
From the above demonstration, we can see that an IAM linked service account does provide the access to the S3 backend, but the mount pod is created with the default service account, c.f.,
https://github.com/juicedata/juicefs-csi-driver/blob/61fd54e1076421b7eed4519f8e50426cf07b44a3/pkg/juicefs/mount/pod.go#L111-L157

To make the mount pod gain the access to S3 backend with IAM role-linked service account, I think we could either inherit the service account of `juicefs-csi-node`, or make it a configurable option.

BTW, the annotations of node service account is not customizable, c.f.,
https://github.com/juicedata/juicefs-csi-driver/blob/61fd54e1076421b7eed4519f8e50426cf07b44a3/charts/juicefs-csi-driver/values.yaml#L60-L65

To link an IAM role, a configurable `.metadata.annotations.eks.amazonaws.com/role-arn` of serviceAccount is mandatory.

Regards.

**Comment by Iced-Sun** [comment]
For anyone who wants to use AWS roles for service accounts, here is the workaround.

1. edit the assume role policy of IAM role (for service accounts):
```json
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Principal": {
        "Federated": "arn:aws-cn:iam::<REDACTED>:oidc-provider/oidc.eks.cn-northwest-1.amazonaws.com.cn/id/<REDACTED>"
      },
      "Action": "sts:AssumeRoleWithWebIdentity",
      "Condition": {
        "StringEquals": {
          "oidc.eks.cn-northwest-1.amazonaws.com.cn/id/<REDACTED>:sub": [
            "system:serviceaccount:juicefs:juicefs-csi-controller-sa",
            "system:serviceaccount:juicefs:juicefs-csi-node-sa",
            // import change here
            "system:serviceaccount:juicefs:default"
          ]
        }
      }
    }
  ]
}
```

2. edit the service account `juicefs/default`, such that
```yaml
apiVersion: v1
kind: ServiceAccount
metadata:
  annotations:
    # import change here
    eks.amazonaws.com/role-arn: arn:aws-cn:iam::<REDACTED>:role/<REDACTED>
  name: default
  namespace: juicefs
...
```

3. edit the cluster role binding `juicefs-csi-node-service-binding`, such that
```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: juicefs-csi-node-service-binding
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: juicefs-csi-external-node-service-role
subjects:
- kind: ServiceAccount
  name: juicefs-csi-node-sa
  namespace: juicefs
  # import change here
- kind: ServiceAccount
  name: default
  namespace: juicefs
```
 
If configured properly, juicefs works just fine without an access key or an instance role.

**Comment by davies** [comment]
That's great, we plan to let mount pod inherit the service account of csi pod, will that be still helpful?

**Comment by Iced-Sun** [comment]
> That's great, we plan to let mount pod inherit the service account of csi pod, will that be still helpful?

Definitely. Official support will be much appreciated as it reduces maintenance burden. We adopt the workaround in https://github.com/juicedata/juicefs-csi-driver/issues/167#issuecomment-970383492 since the migration to the official support will be expected to be minimal and smooth.

Thanks!

**Comment by zwwhdls** [comment]
Hi @Iced-Sun , we have support mount pod inherit the service account of csi pod, pr: https://github.com/juicedata/juicefs-csi-driver/pull/200

You can use nightly image to have a test.

**Comment by a0s** [comment]
Is https://github.com/juicedata/juicefs-csi-driver/issues/167#issuecomment-970383492 is still actual for the current time? Or it all was integrated into helm charts?
